### ISSUE TITLE
Rule for linting `suspend` call in `finally` block

### ISSUE THREAD

**Issue by WildOrangutan** [detekt/detekt#7585]
## Expected Behavior of the rule

Would it make sense adding a rule, that would check, if suspending function is called from `finally` block, without using `NonCancellable`?

For example, in following case, `suspendingCleanup()` will not be executed, if parent coroutine scope was cancelled, while `suspendingWork()` was still running.

```kotlin
try {
    suspendingWork()
} finally {
    suspendingCleanup()
}
```

The rule I'm suggesting, would either enforce using `NonCancellable` coroutine context or require you to suppress it:

```kotlin
try {
    suspendingWork()
} finally {
    withContext(NonCancellable) {
        suspendingCleanup()
    }
}
```

If you want to play around: [Kotlin Playground
](https://pl.kotl.in/WQzzqa6Cz)
## Context

Trying to avoid potential and subtle bugs, when `NonCancellable` is not used.

**Comment by BraisGabin** [comment]
It sound nice.

**Comment by VitalyVPinchuk** [comment]
I'd like to work on the rule if it's still relevant.

**Comment by BraisGabin** [comment]
It is relevant

**Comment by VitalyVPinchuk** [comment]
Should we report all suspend functions in `finally` that aren't wrapped in `NonCancellable`,
or should we attempt to analyze if they are inherently non-cancelable and ignore those? Though it might be challenging to tell if a suspend function is cancelable or not.

**Comment by BraisGabin** [comment]
I think that we should report all of them.

If we could find "inherently non-cancelable" suspend function then we should create another rule to report them.

**Comment by atulgpt** [comment]
Another pattern that is used to perform the non-cancellable task is to use some externalScope or `applicationScope`. We may also avoid flagging those.

**Comment by VitalyVPinchuk** [comment]
> Another pattern that is used to perform the non-cancellable task is to use some externalScope or `applicationScope`. We may also avoid flagging those.

That's correct, but it seems like this might be a concern of a different rule.
Executing `finally` code in a coroutine scope other than the original `try` block feels a bit counterintuitive to me. Is this a common pattern?

**Comment by atulgpt** [comment]
I read this here https://medium.com/androiddevelopers/coroutines-patterns-for-work-that-shouldnt-be-cancelled-e26c40f142ad He suggested a few ways where using application-hosted coroutine scope is also one way. 
` but it seems like this might be a concern of a different rule.` make sense

**Comment by cortinico** [comment]
Closed by https://github.com/detekt/detekt/pull/7664
